docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,519 @@
+    - However, if a user decides to do a mixture of cert stores + local directory then they would need to specify in 2 different sections (cosign plugin config + cert store)
+    - **UPDATE 1/18/24**: We will mirror notation experience with a `verificationKeys` field in the cosign verifier config
+
+## Support Multiple keys for cosign verification
+
+Currently, there is only support for a single key per cosign verifier. 
+
+- Goals:
+  - Cosign verifier should be updated to trust a set of keys and perform verification using that trusted key set
+  - Cosign verifier MUST match the verification logic of Cosign CLI for a single Key
+  - Cosign verifier should be able to operate a loose ('All keys are equally trusted and ONLY one is requred to be used in validation') or strict (ALL keys are trusted and ALL keys must be used in validation) enforcement.
+
+### How does Cosign CLI verification work today?
+
+- Image manifest pushed to same repository as subject (tag schema: `sha256-abc123.sig`)
+- Image manifest is updated every time new signature is added. Each layer is a different signature (possibly from different signing keys)
+```json
+# Sample Cosign OCI Image Manifest
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "config": {
+    "mediaType": "application/vnd.oci.image.config.v1+json",
+    "size": 451,
+    "digest": "sha256:bbacdc3dc4b33c868c6c27192cb1d1bda619c1454abb57ebea806eac13250d3c"
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEQCIGoZazNt1/eEBc08AmcqXY+yNQQJY2ziZPiIopDjHnU5AiAUgeWBy90GgqDdnWpxd1b3lbqzgr2UV2ZyR59eDOyxVg=="
+      }
+    },
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEUCIBVcGGZR6pvJ2SQ8KI0TZY02SLFSSOFJuJph2Qcm8cWZAiEArL+BF5BGRGxELVBeJtAUy8i7L1uuBeUT9fF6TrFYVLQ="
+      }
+    },
+    {
+      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
+      "size": 254,
+      "digest": "sha256:75a45fb043a8abfe52f256cb1a424d8808e886a3f40886228cdfa9b8d75c805d",
+      "annotations": {
+        "dev.cosignproject.cosign/signature": "MEQCIEIsK5yr1kGS1dgRmbudT55oghADdyAOfUwlnayPKAmmAiBTdZkhgHkZdw11tullz7HNOg5O1r5U/j4m2RE39b92Vg=="
+      }
+    }
+  ]
+}
+```
+- Image signature succeeds if AT LEAST one of the signatures verifies true with the corresponding key
+  - Validation occurs on each signature in the manifest in parallel against the SAME key
+  - Cosign only supports validation against a SINGLE key at a time
+
+### Trust Policy
```

Comment:
@yizha1 based on feedback, I came up with a proposal for a trust policy for cosign. Based on how notation does it but has different fields. Implementation would not be done for all features based on current dev scope however I though it'd be useful to introduce concept now so we can add to it as we want to.

Reply:
During the sbom and vulnerability report discussion, we discussed the scenario where artifact and image may requirement different verification key/cert.  Can this be achieved by trust policy scopes? it adding a artifactType filter an option?

Reply:
Thanks for calling it out. That's a good point to consider. Maybe we can add another field `artifactTypeScope` which can act as a filter.